airbyte-config/models/src/main/resources/types/FailureReason.yaml

```diff
@@ -0,0 +1,49 @@
+---
+"$schema": http://json-schema.org/draft-07/schema#
+"$id": https://github.com/airbytehq/airbyte/blob/master/airbyte-config/models/src/main/resources/types/FailureReason.yaml
+title: FailureSummary
+type: object
+required:
+  - failureSource
+  - failureType
+  - timestamp
+additionalProperties: false
+properties:
+  failureSource:
+    description: Indicates where the error originated. If not set, the source of error is not well known.
+    type: string
+    enum:
+      - unknown
+      - source
+      - destination
+      - replicationWorker
+      - persistence
+      - normalization
+      - dbt
+  failureType:
+    description: Categorizes well known errors into types for programmatic handling. If not set, the type of error is not well known.
+    type: string
+    enum:
+      - unknown
+      - permissionError
+      - systemRestart
+      - zombie
+      - manualCancellation
+  internalMessage:
+    description: Human readable failure description for consumption by technical system operators, like Airbyte engineers or OSS users.
+    type: string
+  externalMessage:
+    description: Human readable failure description for presentation in the UI to non-technical users.
+    type: string
+  metadata:
+    description: Key-value pairs of relevant data
+    type: object
+    additionalProperties: true
+  stacktrace:
+    description: Raw stacktrace associated with the failure.
+    type: string
+  retryable:
```

Comment:
@edgao was sharing some feelings on this field that made a lot of sense to me (that was missing in the original spec). That we should have the concepts of: quarantine for the user to fix it, quarantine for airbyte to fix it, retry, unknown. would love to figure out how to add that in this struct some how. @edgao wdyt?

Reply:
Could `failureType` be directly replaced with the error class? And then remove the `retryable` entry. Something along these lines:
* unknown -> unknown
* permissionError -> userError
* systemRestart -> retry (?)
* zombie -> airbyteError (?)
* manualCancellation -> notAnError

I'm kind of torn on whether to keep the more granular `failureType` - on the one hand, it could be interesting to display stats like "in the last 30 days, 60% of your errors were permissions-related" (maybe this could nudge users to have better processes around managing access, or encourage us to build certain features, etc). But this enum feels like it could easily become a very long list that contains every error under the sun, up to and including `targetWidgetDidNotExistBecauseSourceWhizbangWasImproperlyCalibrated`.

Reply:
I think this comment thread from the original tech spec is very relevant: https://docs.google.com/document/d/1vy2iWmiGK5gptLCJVmzYckQj24zFuwF6C_Q-YIlP0is/edit?disco=AAAATDVxx6o .

No worries about contradicting anything! It makes sense that we'd want to iterate on what we want to do with the schema now that we can see how it'll actually interact with and exist in the codebase. I think replacing `retryable (boolean)` with an enum that captures those different resolution paths is a great call.

`failureType` definitely feels hard to keep reasonably scoped. @edgao when you say "error class", do you mean like the actual `Exception.class` value of the original exception?

Reply:
cool. enum makes sense to me!

Reply:
sorry, error class as in unknown/userError/retry/etc. Wanted to avoid overloading the term `failureType` in my comment but am bad at naming >.>

That comment thread does look relevant! Gonna put thoughts here b/c gdoc comments aren't great for longform discussion IMO.

# user intervention
@cgardens does "quarantined for TS" exist in this usecase? To my understanding, we (airbyte) don't really get involved in actually configuring anything, and everything is just configured by the user, so there's no real TS-equivalent. But maybe I'm missing something there.

# more granular failure reasons
I strongly believe the translation from "granular failure reason" (e.g. `permissionError`) to "high-level failure handling strategy" (`userError` + user-friendly message) should live as close to the source of the error (`destination-s3`) as possible. Then the UI/orchestrator can just check that high-level failure type, rather than needing to know about all the different granular failure reasons.

@cgardens what sort of usecases are you envisioning by having the more granular failure reason? I'm seeing it mostly as useful for a human analyst to look at.

What if we changed `failureType` to a string here, but in each individual component is actually an enum? So e.g. the worker might emit one of `zombie|idk|whatever`, and destination-s3 could emit one of `missingPermissions|networkTimeout|bucketDoesNotExist|etc`.

Then it would take some more effort during analysis to map different granular reasons together (e.g. if source-shopify emits a `badOauth` failure, but source-google-ads emits `invalidOauth`, we'd need to tie those together somehow), but at least we'd capture it in a somewhat programmatically-friendly way. They could probably also be standardized to some extent (e.g. any blob store connector could have `bucketDoesNotExist`, any oauth connector could have `invalidClientId`, etc).

(I'm obviously thinking about this in a very connector-centric way. but I think it extends to platform-ish areas fairly intuitively)

Reply:
This can relate to some discussion we had around the new scheduler. The plan was to have 2 type of exceptions thrown by activities: `RetryableException` and `NonRetryableException`. We didn't have time to properly defined what is retryable and what is not. The new scheduler is suppose to be design to handle the scenario you describe and provide the tools to list the quarantine as well as the signal and query method to return the cause and unstuck a sync.

Reply:
@edgao I think you raise some great points that we'll definitely want to consider as we iterate on error/failure handling throughout the platform. I think we'll eventually want to "incentivize" connectors to conform to some standardized form of error/failure reporting, perhaps even defining classes of failures with the expected platform handling in a future version of the protocol.

For now, I think the goal of this PR should primarily be to:
1) make it possible to differentiate between source and destination failures today
2) create a baseline schema for capturing failure metadata that can serve as a starting point for further iteration


I think the `failureSource` solves for goal #1, and we can solve for #2 by supporting an optional `failureType` string field like you suggested, with the expectation that we'll want to iterate towards some sort of standardization before we start deriving much value from it.

Reply:
cool, that makes sense to me :+1: